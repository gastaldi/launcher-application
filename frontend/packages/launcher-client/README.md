# launcher-client [![Build Status](https://travis-ci.com/fabric8-launcher/launcher-client.svg?branch=master)](https://travis-ci.com/fabric8-launcher/launcher-client)
Typescript / Javascript client to communicate with the launcher backends
