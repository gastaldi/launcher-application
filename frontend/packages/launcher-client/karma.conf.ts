module.exports = (config) => {
  config.set({
    frameworks: ['jasmine', 'karma-typescript'],
    files: [
      { pattern: 'src/**/*.ts' }
    ],
    preprocessors: {
      '**/*.ts': ['karma-typescript']
    },
    karmaTypescriptConfig: {
      tsconfig: './tsconfig.json'
    },
    reporters: ['mocha', 'karma-typescript'],
    browsers: ['ChromeHeadless'],
    singleRun: true
  });
};
