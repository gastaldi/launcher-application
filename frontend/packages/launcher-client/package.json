{
  "name": "launcher-client",
  "version": "1.0.0-alpha.0",
  "homepage": "https://github.com/fabric8-launcher/launcher-client",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/fabric8-launcher/launcher-client.git"
  },
  "bugs": {
    "url": "https://github.com/fabric8-launcher/launcher-client/issues"
  },
  "main": "build/index.js",
  "module": "build/esm/index.js",
  "types": "build/types/index.d.ts",
  "engines": {
    "node": ">= 8.9.0",
    "npm": ">= 5.5.0"
  },
  "scripts": {
    "build": "rm -rf ./build && rollup -c",
    "test": "jest",
    "semantic-release": "semantic-release"
  },
  "dependencies": {
    "axios": "0.19.0",
    "injection-js": "2.2.1",
    "lscache": "1.3.0",
    "reflect-metadata": "0.1.13"
  },
  "devDependencies": {
    "@types/axios": "0.14.0",
    "@types/lscache": "1.0.29",
    "@types/node": "11.13.4",
    "axios-mock-adapter": "1.16.0",
    "rollup-plugin-json": "4.0.0",
    "rollup-plugin-strip-code": "0.2.6"
  },
  "eslintConfig": {
    "extends": "react-app"
  }
}
