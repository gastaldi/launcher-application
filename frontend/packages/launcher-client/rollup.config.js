import peerDepsExternal from 'rollup-plugin-peer-deps-external';
import filesize from 'rollup-plugin-filesize';
import tslint from 'rollup-plugin-tslint';
import typescript from 'rollup-plugin-typescript2';
import json from 'rollup-plugin-json';

import pkg from './package.json';

const config = {
    input: 'src/index.ts',
    output: [
        {
            file: pkg.main,
            format: 'cjs',
        },
        {
            file: pkg.module,
            format: 'es',
        },
    ],
    external: [
        ...Object.keys(pkg.dependencies || {}),
        ...Object.keys(pkg.peerDependencies || {}),
    ],
    plugins: [
        peerDepsExternal(),
        tslint({
            include: './src/**/*.tsx?'
        }),
        typescript({
            clean: true,
            typescript: require('typescript'),
            verbosity: 0,
            useTsconfigDeclarationDir: true
        }),
        json(),
        // FIXME find another solution to remove the mock (that make it possible to choose from the consumer side)
        /**
        isProduction && stripCode({
            start_comment: 'test-code',
            end_comment: 'end-test-code'
        }),**/
        filesize(),
    ],
};

export default config;
