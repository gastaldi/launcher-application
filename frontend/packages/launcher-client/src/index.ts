/*
 * Public API
 */

export * from './client/launcher.client.factory';
export * from './client/launcher.client';
export * from './client/types';
export * from './client/helpers/launchers';
export * from './client/helpers/mock-helpers';
export * from './client/helpers/preconditions';
