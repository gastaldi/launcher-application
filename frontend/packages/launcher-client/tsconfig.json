{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "rootDir": "src",
    "declaration": true,
    "declarationDir": "build/types"
  },
  "include": [
    "src"
  ]
}
