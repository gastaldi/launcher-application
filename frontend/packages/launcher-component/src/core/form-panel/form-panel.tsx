import * as React from 'react';
import { useState } from 'react';
import { Button, Toolbar, ToolbarGroup } from '@patternfly/react-core';
import { InputProps } from '../types';
import { Separator } from '../stuff';

interface FormPanelProps<T> {
  id?: string;
  initialValue: T;

  validator?(value: T): boolean;

  children(inputProps: InputProps<T>): any;

  onSave?(value: T);

  onCancel?();
}

export function FormPanel<T>(props: FormPanelProps<T>) {
  const [value, onChange] = useState<T>(props.initialValue);

  const onSave = () => {
    if (props.onSave) {
      props.onSave(value);
    }
  };

  const onCancel = () => {
    onChange(props.initialValue);
    if (props.onCancel) {
      props.onCancel();
    }
  };
  const id = props.id || 'form';
  return (
    <div className="form-panel" style={{padding: '20px'}}>
      {props.children({value, onChange})}
      <Separator/>
      <Toolbar>
        <ToolbarGroup>
          <Button
            variant="primary"
            aria-label={'Save ' + id}
            onClick={onSave}
            isDisabled={props.validator && !props.validator(value)}
          >
            Save
          </Button>
        </ToolbarGroup>
        <ToolbarGroup>
          <Button variant="secondary" aria-label={'Cancel ' + id} onClick={onCancel}>Cancel</Button>
        </ToolbarGroup>
      </Toolbar>
    </div>
  );
}
