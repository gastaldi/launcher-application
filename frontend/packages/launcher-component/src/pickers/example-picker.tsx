import * as React from 'react';
import {
  DataList,
  DataListCell,
  DataListContent,
  DataListItem,
  Radio,
  Title,
  FormSelect,
  FormSelectOption,
  Button,
  Split,
  SplitItem
} from '@patternfly/react-core';
import { ExampleRuntime, ExampleMission } from 'launcher-client';
import { InputProps, Picker } from '../core/types';
import _ from 'lodash';

export interface ExamplePickerValue {
  missions?: ExampleMission[];
  missionId?: string;
  runtimeId?: string;
  versionId?: string;
  downloadOnly?: boolean;
}

interface ExamplePickerProps extends InputProps<ExamplePickerValue> {
  missions?: ExampleMission[];
}

function isDownloadOnly(version) {
  return (!!version && (version.metadata.runsOn.indexOf('none') !== -1)) || false;
}

export const ExamplePicker: Picker<ExamplePickerProps, ExamplePickerValue> = {
  checkCompletion: value => !!value.missionId && !!value.runtimeId && !!value.versionId,
  Element: props => {
    // @ts-ignore
    const runtimesMap = _.keyBy(_.flatten(_.map(props.missions, 'runtime')), 'id');
    const missions = !props.value.missions ? props.missions : props.value.missions;
    const filterCatalog = (runtime) => {
      const filteredMissions = _.cloneDeep(props.missions);
      filteredMissions!.map(m => m.runtime = (m.runtime as ExampleRuntime[]).filter(r => r.id === runtime));
      props.onChange({
        runtimeId: runtime,
        missions: filteredMissions!.filter(m => (m.runtime as ExampleRuntime[]).find(r => !!r))
      });
    };
    return (
      <React.Fragment>
        <Split>
          <SplitItem isMain>
            Already decided what runtime you want to use? Select it here:
          </SplitItem>
          <SplitItem isMain={false}>
            <Button variant="tertiary" onClick={() => props.onChange({...props.value, missions: undefined, runtimeId: ''})}>Any</Button>
            {_.map(runtimesMap).map((runtime, index) => (
              <Button
                style={{marginLeft: '5px'}}
                variant="tertiary"
                key={index}
                onClick={() => filterCatalog(runtime.id)}
                isDisabled={props.value.runtimeId === runtime.id && !!props.value.missions}
              >
                {runtime.name}
              </Button>
            ))}
          </SplitItem>
        </Split>
        <DataList aria-label="Choose an example" style={{marginTop: '20px'}}>
          {
            missions!.map((mission, i) => {
              const isSelected = props.value.missionId === mission.id;
              const onChange = (runtimeId = props.value.runtimeId, versionId?) => {
                const runtime = (mission.runtime! as any).find(r => r.id === runtimeId);
                if (runtimeId) {
                  if (!runtime || !runtime.versions || runtime.versions.lenght <= 0) {
                    runtimeId = undefined;
                  } else if(!versionId) {
                    versionId = runtime.versions[0].id;
                  }
                }
                const version = runtime && runtime.versions.find(v => v.id === versionId);
                props.onChange({
                  ...props.value,
                  missionId: mission.id,
                  runtimeId,
                  versionId,
                  downloadOnly: isDownloadOnly(version)
                });
              };
              const selectedRuntime = (mission.runtime! as any).find(r => r.id === props.value.runtimeId);
              return (
                <DataListItem
                  isExpanded={isSelected}
                  aria-labelledby={mission.name}
                  value={mission.id}
                  key={i}
                  style={{cursor: 'pointer'}}
                >
                  <DataListCell width={1} style={{flex: 'none'}}>
                    <Radio
                      aria-label={`Choose ${mission.id} as mission`}
                      value={mission.id}
                      checked={isSelected}
                      onChange={() => onChange()}
                      name="mission"
                      id={`radio-choose-${mission.id}-as-mission`}
                    />
                  </DataListCell>
                  <DataListCell width={1} onClick={() => onChange()}><Title size="lg">{mission.name}</Title></DataListCell>
                  <DataListCell width={2} onClick={() => onChange()}>{mission.description}</DataListCell>
                  <DataListContent aria-label={`Detail for ${mission.id}`} isHidden={!isSelected}>
                    <FormSelect
                      isDisabled={mission.runtime!.length === 1}
                      id={`${mission.id}-runtime-select`}
                      value={props.value.runtimeId || ''}
                      onChange={r => onChange(r)}
                      aria-label="Select Runtime"
                    >
                      {mission.runtime!.length !== 1 && <FormSelectOption
                          value=""
                          label="Select a Runtime"
                      />}
                      {mission.runtime!.map((runtime, index) => (
                        <FormSelectOption
                          key={index}
                          value={runtime.id}
                          label={runtime.name}
                        />
                      ))
                      }
                    </FormSelect>
                    {!!selectedRuntime &&
                    <FormSelect
                        id={`${mission.id}-version-select`}
                        value={props.value.versionId}
                        onChange={value => onChange(selectedRuntime.id, value)}
                        aria-label="Select Version"
                    >
                      {selectedRuntime.versions.map((version, index) => (
                        <FormSelectOption
                          key={index}
                          value={version.id}
                          label={version.name}
                        />
                      ))
                      }
                    </FormSelect>
                    }
                  </DataListContent>
                </DataListItem>
              );
            })
          }
        </DataList>
      </React.Fragment>
    );
  }

};
