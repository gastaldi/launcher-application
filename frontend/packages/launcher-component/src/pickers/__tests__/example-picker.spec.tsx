import * as React from 'react';
import { cleanup, render, fireEvent } from 'react-testing-library';
import { ExamplePicker } from '../example-picker';
import { ExampleMission } from 'launcher-client';
import { FormPanel } from '../../core/form-panel/form-panel';

afterEach(cleanup);

describe('<ExamplePicker />', () => {
  it('renders the ExamplePicker correctly', () => {
    const comp = render(<ExamplePicker.Element missions={missions} value={{}} onChange={() => {}}/>);
    expect(comp.asFragment()).toMatchSnapshot();
  });

  it('show the version dropdown', () => {
    const handleSave = jest.fn();
    const comp = render(
      <FormPanel
        initialValue={{}}
        validator={ExamplePicker.checkCompletion}
        onSave={handleSave}
        onCancel={() => {}}
      >
        {(inputProps) => (<ExamplePicker.Element missions={missions} {...inputProps}/>)}
      </FormPanel>
    );
    const missionRadio = comp.getByLabelText('Choose circuit-breaker as mission');
    fireEvent.click(missionRadio);
    expect(comp.getByLabelText('Select Runtime')).toBeDefined();
    fireEvent.change(comp.getByLabelText('Select Runtime'), { target: { value: 'nodejs' } });
    expect(comp.getByDisplayValue('10.x (Community)')).toBeDefined();
    fireEvent.change(comp.getByLabelText('Select Version'), { target: { value: 'community' } });
    fireEvent.click(comp.getByText('Save'));
    expect(handleSave).toHaveBeenCalledTimes(1);
    expect(handleSave.mock.calls[0][0]).toMatchSnapshot('Saved value');
    expect(comp.asFragment()).toMatchSnapshot('Component');
  });

  it('should not be downloadOnly with runsOn "local"', () => {
    downloadOnlyTest(['local']);
  });

  it('should not be downloadOnly with runsOn "!starter"', () => {
    downloadOnlyTest(['!starter']);
  });

  it('should be downloadOnly for runsOn "none"', () => {
    downloadOnlyTest('none');
  });
});

const downloadOnlyTest = runsOn => {
  const handleSave = jest.fn();
  const comp = render(
    <FormPanel
      initialValue={{}}
      validator={ExamplePicker.checkCompletion}
      onSave={handleSave}
      onCancel={() => {}}
    >
      {(inputProps) => (<ExamplePicker.Element missions={missions} {...inputProps}/>)}
    </FormPanel>
  );
  missions[0].runtime![0].versions[0].metadata.runsOn = runsOn;
  const missionRadio = comp.getByLabelText('Choose circuit-breaker as mission');
  fireEvent.click(missionRadio);
  fireEvent.change(comp.getByLabelText('Select Runtime'), { target: { value: 'nodejs' } });
  fireEvent.change(comp.getByLabelText('Select Version'), { target: { value: 'community' } });
  fireEvent.click(comp.getByText('Save'));
  expect(handleSave).toHaveBeenCalledTimes(1);
  expect(handleSave.mock.calls[0][0]).toMatchSnapshot(`Value with "${runsOn}"`);
};

const missions: ExampleMission[] = [
  {id: 'circuit-breaker', description: 'circuit-breaker description', name: 'Circuit Breaker', metadata: {}, runtime: [
    {id: 'nodejs', name: 'Node', icon: '', versions: [
      {id: 'community', name: '10.x (Community)', metadata: { runsOn: 'openshift'}}
    ]}
  ]}
];
