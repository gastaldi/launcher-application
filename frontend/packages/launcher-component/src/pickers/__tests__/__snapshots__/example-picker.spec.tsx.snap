// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExamplePicker /> renders the ExamplePicker correctly 1`] = `
<DocumentFragment>
  <div
    class="pf-l-split"
  >
    <div
      class="pf-l-split__item pf-m-main"
    >
      Already decided what runtime you want to use? Select it here:
    </div>
    <div
      class="pf-l-split__item"
    >
      <button
        class="pf-c-button pf-m-tertiary"
        type="button"
      >
        Any
      </button>
      <button
        class="pf-c-button pf-m-tertiary"
        style="margin-left: 5px;"
        type="button"
      >
        Node
      </button>
    </div>
  </div>
  <ul
    aria-label="Choose an example"
    class="pf-c-data-list"
    role="list"
    style="margin-top: 20px;"
  >
    <li
      aria-labelledby="Circuit Breaker"
      class="pf-c-data-list__item"
      style="cursor: pointer;"
      value="circuit-breaker"
    >
      <div
        class="pf-c-data-list__cell"
        style="flex: 0 0 auto;"
      >
        <div
          class="pf-c-radio"
        >
          <input
            aria-invalid="false"
            aria-label="Choose circuit-breaker as mission"
            class="pf-c-radio__input"
            id="radio-choose-circuit-breaker-as-mission"
            name="mission"
            type="radio"
            value="circuit-breaker"
          />
        </div>
      </div>
      <div
        class="pf-c-data-list__cell"
      >
        <h1
          class="pf-c-title pf-m-lg"
        >
          Circuit Breaker
        </h1>
      </div>
      <div
        class="pf-c-data-list__cell pf-m-flex-2"
      >
        circuit-breaker description
      </div>
      <section
        aria-label="Detail for circuit-breaker"
        class="pf-c-data-list__expandable-content"
        hidden=""
        id=""
      >
        <select
          aria-invalid="false"
          aria-label="Select Runtime"
          class="pf-c-form-control"
          disabled=""
          id="circuit-breaker-runtime-select"
        >
          <option
            class=""
            value="nodejs"
          >
            Node
          </option>
        </select>
      </section>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`<ExamplePicker /> should be downloadOnly for runsOn "none": Value with "none" 1`] = `
Object {
  "downloadOnly": true,
  "missionId": "circuit-breaker",
  "runtimeId": "nodejs",
  "versionId": "community",
}
`;

exports[`<ExamplePicker /> should not be downloadOnly with runsOn "!starter": Value with "!starter" 1`] = `
Object {
  "downloadOnly": false,
  "missionId": "circuit-breaker",
  "runtimeId": "nodejs",
  "versionId": "community",
}
`;

exports[`<ExamplePicker /> should not be downloadOnly with runsOn "local": Value with "local" 1`] = `
Object {
  "downloadOnly": false,
  "missionId": "circuit-breaker",
  "runtimeId": "nodejs",
  "versionId": "community",
}
`;

exports[`<ExamplePicker /> show the version dropdown: Component 1`] = `
<DocumentFragment>
  <div
    class="form-panel"
    style="padding: 20px;"
  >
    <div
      class="pf-l-split"
    >
      <div
        class="pf-l-split__item pf-m-main"
      >
        Already decided what runtime you want to use? Select it here:
      </div>
      <div
        class="pf-l-split__item"
      >
        <button
          class="pf-c-button pf-m-tertiary"
          type="button"
        >
          Any
        </button>
        <button
          class="pf-c-button pf-m-tertiary"
          style="margin-left: 5px;"
          type="button"
        >
          Node
        </button>
      </div>
    </div>
    <ul
      aria-label="Choose an example"
      class="pf-c-data-list"
      role="list"
      style="margin-top: 20px;"
    >
      <li
        aria-labelledby="Circuit Breaker"
        class="pf-c-data-list__item pf-m-expanded"
        style="cursor: pointer;"
        value="circuit-breaker"
      >
        <div
          class="pf-c-data-list__cell"
          style="flex: 0 0 auto;"
        >
          <div
            class="pf-c-radio"
          >
            <input
              aria-invalid="false"
              aria-label="Choose circuit-breaker as mission"
              class="pf-c-radio__input"
              id="radio-choose-circuit-breaker-as-mission"
              name="mission"
              type="radio"
              value="circuit-breaker"
            />
          </div>
        </div>
        <div
          class="pf-c-data-list__cell"
        >
          <h1
            class="pf-c-title pf-m-lg"
          >
            Circuit Breaker
          </h1>
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-2"
        >
          circuit-breaker description
        </div>
        <section
          aria-label="Detail for circuit-breaker"
          class="pf-c-data-list__expandable-content"
          id=""
        >
          <select
            aria-invalid="false"
            aria-label="Select Runtime"
            class="pf-c-form-control"
            disabled=""
            id="circuit-breaker-runtime-select"
          >
            <option
              class=""
              value="nodejs"
            >
              Node
            </option>
          </select>
          <select
            aria-invalid="false"
            aria-label="Select Version"
            class="pf-c-form-control"
            id="circuit-breaker-version-select"
          >
            <option
              class=""
              value="community"
            >
              10.x (Community)
            </option>
          </select>
        </section>
      </li>
    </ul>
    <hr
      class="separator"
    />
    <div
      class="pf-l-toolbar"
    >
      <div
        class="pf-l-toolbar__group"
      >
        <button
          aria-label="Save form"
          class="pf-c-button pf-m-primary"
          type="button"
        >
          Save
        </button>
      </div>
      <div
        class="pf-l-toolbar__group"
      >
        <button
          aria-label="Cancel form"
          class="pf-c-button pf-m-secondary"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ExamplePicker /> show the version dropdown: Saved value 1`] = `
Object {
  "downloadOnly": false,
  "missionId": "circuit-breaker",
  "runtimeId": "nodejs",
  "versionId": "community",
}
`;
